src/Storages/MergeTree/IMergeTreeDataPart.cpp

```diff
@@ -1643,6 +1643,12 @@ void IMergeTreeDataPart::remove()
             return CanRemoveDescription{.can_remove_anything = true, .files_not_to_remove = {} };
         }
 
+        if (getState() == MergeTreeDataPartState::Temporary)
+        {
+            LOG_TRACE(storage.log, "Part {} in temporary state can be removed without unlocking shared state", name);
+            return CanRemoveDescription{.can_remove_anything = true, .files_not_to_remove = {} };
```

Comment:
This is completely wrong and not what I intended to do.
So. Now  we do not call canRemovePart here, but remove all blobs from S3. If fetch has fail at some point to rename part from tmp-fetch_ to the destination dir, part destructor is removing shared data from S3 without checking refCounts. Intention was opposit, remove from S3 nothing, just clear the local disk.